Abstract out the network IO

sechkova opened the pull request:
**Description of the changes being introduced by the pull request**:

Based on the discussion in #1213 and mainly the idea of @jku, implement an abstract `FetcherInterface` calss and
a concrete `RequestsFetcher` class .

The idea is to extract the network IO related code out of `tuf.download` and put it in `RequestsFetcher` class which is also the default `FetcherInterface` implementation which TUF uses if not given an external one.

`RequestsFetcher` implements only one public method '`fetch`' which returns a bytes iterator. All file handling and verification
logic remains in `tuf.download` which is now wrapper in a `FileDownload` class for convenience.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
Add FetcherInterface class

The new class FetcherInterface defines an interface for
abstract network download which can be implemented for a
variety of network libraries and configurations.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Move network IO logic to RequestsFetcher

Abstract the network IO. Move the network operations from
tuf.download to the RequestsFercher class which is TUF's
implementation of the abstract FetcherInterface.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add fetcher as parameter to Updater class

Initialize Updater with an external implementation of
FetcherInterface. If not provided, tuf.fetcher.RequestsFetcher
is used.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Update tests importing tuf.download

Pass RequestsFetcher object to tuf.download functions.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add test_fetcher

Add unit test for requests_fetcher.py

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Implement RequestsFetcher class

A concrete implementation of FetcherInterface based on
the Requests library.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
The new GitHub actions have been harsh on me :)) Will review the errors after the holiday break.

sechkova commented:
> The new GitHub actions have been harsh on me :)) Will review the errors after the holiday break.

Fixed the linter issues, however according to Windows "Download was too slow." :roll_eyes:

sechkova commented:
Fixed a small issue and force-pushed but (un)fortunately I see no errors while testing on a local windows VM.

sechkova committed:
Take out connection time from download speed calculation

- Update RequestsFetcher.fetch to return a generator object.
- Update _download_file to skip connection time when calculating
average download speed.
- Write chunk to temp file before exiting the fetcher loop
on error.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
First thing, apologize for the spam and using the CI as a test server but I could not reproduce the Windows tests failure locally.

The problem can be summarized in two parts:

1. Slow 'localhost' lookup on Windows #1257
2. A change in this PR which adds the time to connect to the server to the calculation of `average_download_speed` for the first downloaded chunk which raises `SlowRetreivalError` in the case of 1.

While 1. can quickly be fixed as suggested by @jku in #1257, 2. is related to the network abstraction api which may need some discussion. The main point is, given all the uncertainty around protection against slow retrieval attack and its recent removal from the specification, where should we implement the slow retrieval check:

1. Keep the fetcher unaware of it and implement the download speed calculations in its caller function.
   - has the obvious pros that fetcher implementation is unaware of it and the check is performed by TUF itself
   - cons: incorrectly adds connection time to the actual chunk download time
2. Keep the fetcher unaware of it but make fetcher API more complicated by splitting session establishment/connection and downloading data
    - pros: solves the timing issue
    - cons: undesirably complicated api
3. Move the download speed calculation and slow retrieval checks as part of TUF's fetcher implementation
    - pros: replicates current behaviour
    - cons: different fetcher implementations can(and most likely will) totally ignore the slow retrieval check

My choice is in favour of point 3 but I would like to hear more opinions before going ahead.

lukpueh commented:
Thanks for the detailed comment, @sechkova (and of course for the patch, which I haven't looked at yet). Here are my thoughts:
- I'd rule out (2). This attack, which is no longer part of the spec (https://github.com/theupdateframework/specification/pull/111), does not seem to justify a more complicated API.
- I think (3) is perfectly fine, maybe we can add a big disclaimer to the API docs that says that only the default fetcher provides this protection, and any custom implementation must do it themselves.
- Regardless, wouldn't it be possible to achieve (1) by granting some grace period in the beginning, where the fetcher may drop below the rate? AFAIU a successful slow-retrieval attack requires quite some time, so if we start checking the average rate only after enough time has passed to likely smooth out the spike caused by connection/session establishment, we are probably able to detect the attack early enough.

trishankatdatadog commented:
I agree with @lukpueh: (1) makes more sense to me, depending on the philosophical motivation for the fetcher. Are we simply trying to fetch bytes or do something more?

jku commented:
I lean towards option 3 (mostly because I do not believe our slow retrieval protection works without false positives -- Teodoras test failures are a strange example of this IMO -- and I would like to not have these false positives in pip)

> wouldn't it be possible to achieve (1) by granting some grace period in the beginning, where the fetcher may drop below the rate? AFAIU a successful slow-retrieval attack requires quite some time, so if we start checking the average rate only after enough time has passed to likely smooth out the spike caused by connection/session establishment, we are probably able to detect the attack early enough.

I originally suggested the same thing, as this might also help with the massive bias that the current implementation has. Example of the bias: Imagine a download of 10 chunks where 9 chunks transfer at "normal internet speeds" but one is slow:  
* if the slow chunk is the first one the download will be canceled if speed is < 50B/s
* if the slow chunk is the last one the download will happily succeed even if the chunk speed is 1 B/s (a very long time later but it will succeed)

However, solving the above issue issue properly is no doubt complex -- probably the best we could do is special case the first chunk handling, just giving it a few extra seconds... I don't see why this is better than letting the fetcher handle it.

jku commented:
Oh additional reason for choice 3 (that Teodora kind of mentioned): the patch will probably be cleaner as it will be easier to just move existing code from download.py to Fetcher instead of developing similar but not quite same code.

jku commented:
> I agree with @lukpueh: (1) makes more sense to me, depending on the philosophical motivation for the fetcher. Are we simply trying to fetch bytes or do something more?

The issue is that the first call to the fetch() generator also creates the connection because there is no separate initialization function... now that I said it aloud, I think it has to be possible to initialize the generator separately. I'll look this up. Thanks.

joshuagl commented:
I am in favour of option three, both because we don't have a specification for slow retrieval attacks, and because we are not confident about the effectiveness of the current implementation of slow retrieval attacks.

> First thing, apologize for the spam and using the CI as a test server but I could not reproduce the Windows tests failure locally.

I didn't receive any spam about CI failures, so I think no need to apologise here 😄

sechkova committed:
Avoid 'localhost' lookup in tests

On Windows (Github Actions) the lookup for 'localhost' takes 1 second.
This is because:
- Windows retries connect() with a timeout
- the machine has IPv6 and IPv4 but Testserver only binds the port on IPv4
- the test clients connect to 'localhost'

Since socketserver.TCPServer does not seem to support IPv6 before 3.8,
just replace 'localhost' with '127.0.0.1' in client-side URLs.

See #1257

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Thanks all for giving your opinion and especially @jku for the ideas, most of which I think are included in the latest changes. 
Feel free to review :wink:

joshuagl commented on tuf/fetcher.py:
```diff
@@ -113,12 +113,11 @@ def fetch(self, url, required_length):
-
-      # Check response status.
-      response.raise_for_status()
+    response = session.get(url, stream=True, timeout=self.SOCKET_TIMEOUT)
```
As there's no longer a context manager in use for `response`, do we need to call some cleanup method?

jku replied:
also maybe double check what happens when connect times out. I expect that's a ReadTimeoutError which should be handled

sechkova replied:
Did some cleanup in 77900af.

Connect time out will raise an exception that will be captured in `NoWorkingMirrorError`. This is the same as the current behaviour.

jku commented on tuf/fetcher.py:
```diff
@@ -0,0 +1,201 @@
+
+
+
+class FetcherSettings():
```
Can you explain the construct? I can see it should work but what is the advantage of a separate FetcherSettings and RequestsFetcher?

sechkova replied:
The idea was to separate `FetcherSettings` so that other `FetcherInterface` implementations can reuse them (by inheriting) or replace them with their own. But given that we cannot remove them from `tuf.settings` just like that, I dropped this change too.

jku commented on tuf/settings.py:
```diff
@@ -74,12 +74,6 @@
 # download Targets metadata.
 DEFAULT_TARGETS_REQUIRED_LENGTH = 5000000 #bytes
 
-# Set a timeout value in seconds (float) for non-blocking socket operations.
-SOCKET_TIMEOUT = 4 #seconds
-
-# The maximum chunk of data, in bytes, we would download in every round.
-CHUNK_SIZE = 400000 #bytes
-
```
These are settings that unfortunately maybe used by someone so I don't think we should just silently delete them without deprecation.

I wonder if the new implementation should still use these as the default values for its properties... and possibly mark these as deprecated, explaining that these values are now fetcher specific and that the default RequestsFetcher exposes the same properties (but that those properties might not be considered API(?) and those changing them should consider making their own fetcher)

sechkova replied:
I totally forgot about `settings` being part of the API. I dropped the commit and just used `tuf.settings` in `RequestsFetcher`. 

I suggest we revise `tuf.settings` in general and then apply the decisions here.

jku commented on tuf/client/updater.py:
```diff
@@ -659,6 +659,10 @@ def __init__(self, repository_name, repository_mirrors):
                                           'targets_path': 'targets',
                                           'confined_target_dirs': ['']}}
 
+      fetcher:
+        A conctrete 'FetcherInterface' implementation. If not provided,
+        tuf.fetcher.RequestsFetcher is used.
```
typo on concrete. Also it's lacking a sentence on what fetcher does.

sechkova replied:
Fixed in b0e8acb

jku reviewed (commented):
I like it except for the fact that it rewrites download.py completely :) I understand it's mostly an indent change but I was really hoping that wasn't needed.

I have not yet reviewed download.py changes -- I will do that but I'll leave an early review for this question: _What compromises would we need to make to keep download.py changes to a minimum?_

sechkova commented:
Reworked some changes after discussion with @jku. Download.py should be more appealing for a review now!

MVrachev commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,119 @@
+    self.temp_file.seek(0)
+    temp_file_data = self.temp_file.read().decode('utf-8')
+    self.assertEqual(self.file_contents, temp_file_data)
+    self.assertEqual(self.file_length, len(temp_file_data))
```
Do you need to check that when you know that the `self.file_contents` is equal to `temp_file_data` from the previous `assertEqual`?

sechkova replied:
Nope. Thanks for noting, seems like I'll have to take another look on the test cases.

MVrachev commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,119 @@
+
+  # Test incorrect URL parsing
+  def test_url_parsing(self):
+   with self.assertRaises(tuf.exceptions.URLParsingError) as cm:
```
What exactly will happen if this test fails? What about when it succeeds?  
I just haven't seen the usage of a context manager with an exception...

jku replied:
```
with self.assertRaises(ExpectedError):
  function_under_test(argument1, argument2)
```
is functionally the same as
```
self.assertRaises(ExpectedError, function_under_test, argument1, argument2)
```
but more readable especially with longer names and more arguments, and also more extensible (as you can have more than one line in the block as test_url_parsing() does).

The `as cm` seems unneeded (but harmless) in this case as it's not used.

sechkova replied:
> but more readable especially with longer names and more arguments, and also more extensible (as you can have more than one line in the block as test_url_parsing() does).

This is what I've meant when writing it but I guess I had plans to use `cm` that didn't happen ...

jku commented on tuf/download.py:
```diff
@@ -210,174 +191,46 @@ def _download_file(url, required_length, STRICT_REQUIRED_LENGTH=True):
-
-      if number_of_bytes_received == required_length:
-        break
+    chunks = fetcher.fetch(url, required_length)
+    start_time = timeit.default_timer()
+    for chunk in chunks:
```
This is not totally Fetcher related but maybe should be fixed now that we're making changes: I think we never test a download with more than 1 chunk. This is not ideal especially now that a download loop exists in both fetcher and download.py.

Should one or both cases ensure that multiple chunks are tested, maybe by lowering the chunk size to something small or creating a large target file dynamically? Small chunk size is probably a better option (although artificially small chunk size might lead to SlowRetrieval checks being even more unreliable...)

jku replied:
this is handled by Teodoras "Test downloading data in more than one chunk" commit

jku commented on tuf/fetcher.py:
```diff
@@ -0,0 +1,188 @@
+    # Always set the timeout. This timeout value is interpreted by requests as:
+    #  - connect timeout (max delay before first byte is received)
+    #  - read (gap) timeout (max delay between bytes received)
+    response = session.get(url, stream=True,
+        timeout=tuf.settings.SOCKET_TIMEOUT)
```
At some point I think we want to do something about the settings here (like deprecate the ones in settings module and make them just properties in the fetcher implementation) -- but I think this is the correct approach for this PR

sechkova replied:
I was sure that `tuf.settings` has been discussed in terms of TUF 1.0.0 plans but I couldn't find anything documented. Will file an issue about it if nothing comes up.

joshuagl replied:
We have an [issue in securesystemslib](https://github.com/secure-systems-lab/securesystemslib/issues/219), but not a corresponding issue in tuf. We should file one as part of the refactor project and milestones.

sechkova replied:
Here it is  https://github.com/theupdateframework/tuf/issues/1273

jku commented on tuf/fetcher.py:
```diff
@@ -0,0 +1,188 @@
+
+          yield data
+
+          if bytes_received == required_length:
```
this is copied from old implementation but...
the defensive version of this check would be `if bytes_received >= required_length:` -- there should be no harm in using it. download.py already ensures that the length matches so this would just make sure the response is closed etc.

jku commented on tuf/fetcher.py:
```diff
@@ -0,0 +1,188 @@
+    # Check response status.
+    response.raise_for_status()
+
+    def chunks():
```
could add a comment describing the trick here: chunks is the actual iterator that gets returned from fetch() after response object has been initialized (and most importantly after connect() has happened)

jku commented on tuf/fetcher.py:
```diff
@@ -0,0 +1,188 @@
+      self._sessions[session_index] = session
+
+      # Attach some default headers to every Session.
+      requests_user_agent = session.headers['User-Agent']
+      # Follows the RFC: https://tools.ietf.org/html/rfc7231#section-5.5.3
+      tuf_user_agent = 'tuf/' + tuf.__version__ + ' ' + requests_user_agent
+      session.headers.update({
+          # Tell the server not to compress or modify anything.
+          # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding#Directives
+          'Accept-Encoding': 'identity',
+          # The TUF user agent.
+          'User-Agent': tuf_user_agent})
```
user-agent could be a property of the fetcher -- but this makes sense to do in a later PR

sechkova replied:
I think for a later PR it is worth considering going even further and providing a way to [customise requests.Session](https://github.com/theupdateframework/tuf/issues/1188).

joshuagl replied:
Let's be sure to file an issue to track this.

sechkova replied:
Sorry, my link to the existing issue above was not working.
Do we need another one more specific to this PR's changes? I don't want to open an issue about something which has not yet landed in the develop branch.

joshuagl replied:
We should probably update #1188 with recommendations for how to achieve the request within the new fetcher world, once this PR has landed.

jku reviewed (commented):
This is very much what I imagined it would be, thanks! I'm documenting my understanding here, do fix if there are mistakes:
* the issue with connect-time being counted in the first chunk download time was solved by separating the iterator initialization and first chunk download: now it operates just like the current code in this regard
* practical results should be identical to current functionality (and no tests were changed except for interface changes mentioned below and the unrelated localhost bugfix)
* The effects on TUF API are:
  - updater.py API is backwards-compatible (but has a new optional fetcher argument)
  - fetcher.py was added: FetcherInterface is now part of client API. A default implementation is provided and other implementations can be used
  - download.py was considered an implementation detail of updater.py and so download.py interface was modified: some functions now have a new required argument. The new argument could be made optional but that would lead to download.py (in the default case) to always instantiate a new fetcher, and thus new session which is not optimal. I guess it could also keep a "global default fetcher"  much like it now has a "global _sessions" but that seems like a worse idea
  - TUF no longer strictly depends on Requests (but the default fetcher does)

I left some comments in code. I've not had time to integrate this into my pip branch yet but from my POV this looks good to go (with the comments addressed).

sechkova committed:
Apply a defensive data length check in fetch()

Use '>=' as the defensive version of the equality check.

Add a comment describing the need of a chunks() generator.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Test downloading data in more than one chunk

Add test cases to test_fetcher and test_download that
decrease default chunk size and download data in more
than one chunk.

Small code-style improvements.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
> This is very much what I imagined it would be, thanks! I'm documenting my understanding here, do fix if there are mistakes:
> 

Thanks for the summary, @jku, I couldn't have described it better!
Comments are addressed in two new commits.

jku commented:
Thank you.

With regards to the new tests: I'm not against running 2MB of data through the code but writing so much to disk per test makes me a bit uncomfortable if there are alternatives... The other observation I have is that even the fetcher test never actually confirms that multiple chunks were used: we just assume requests is working like we expect. Maybe this is me being a bit paranoid. Thoughts on these?

jku commented:
Other discussion items (Teodora already has good ideas for these but I'm putting the questions out there for others to chime in): 

1. This PR currently suffers from the common issue of class & module name "conflict":
```
from tuf import fetcher
fetcher = fetcher.Fetcher() # oops name clash with variable and module
```
Basically the "logical" name for an instance of the class happens to be the name of the module.  This is not horrible and Updater is similar but if we can avoid it, it would be nice

2. The "API boundary" is as unclear as ever: would be nice if client developers had a clear boundary (e.g. "only files inside client/ are considered public client API")

sechkova commented:
> With regards to the new tests: I'm not against running 2MB of data through the code but writing so much to disk per test makes me a bit uncomfortable if there are alternatives... The other observation I have is that even the fetcher test never actually confirms that multiple chunks were used: we just assume requests is working like we expect. Maybe this is me being a bit paranoid. Thoughts on these?

I amended the last commit which adds tests with more than one chunk download (https://github.com/theupdateframework/tuf/pull/1250/commits/1e4b944310845974709f61ac84d6722159e56d9c). Instead of writing big files on disk, default chunk size is decreased. Let me know what do you think. 
I have doubts about the new test case in `test_download` since there is no visibility on the actual chunks from the `download_file` output so the test only confirms that file is downloaded without errors. It is harmless but is it also useless?

jku commented:
tests look good to me.

I found one real issue: Updater still uses HTTPError from requests (because it wants to know if error is 404 or 403 to decide if root download should error our or not). I think we need to define a HTTPError of our own and require the Fetcher to wrap any http errors in that. A bit annoying but easy to implement in the fetcher: let's experiment with this a bit before committing to a solution.

Commit:
Add an HTTP error for Fetcher interface

A custom error is required so that updater is able to special case
403 & 404 status codes.

Rewrite the test case a bit to be more readable.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
> I found one real issue: Updater still uses HTTPError from requests (because it wants to know if error is 404 or 403 to decide if root download should error our or not). I think we need to define a HTTPError of our own and require the Fetcher to wrap any http errors in that. A bit annoying but easy to implement in the fetcher: let's experiment with this a bit before committing to a solution.

I've tested this and the above is correct. The interface has to require fetch() to throw a specific exception on HTTP errors, otherwise we can't check for 403 & 404 in updater. The actual implementation work is minimal. I'll update this PR (as Teodora is out-of-office this week) hopefully tomorrow.

trishankatdatadog commented:
@jku Ah, right, I totally forgot about that, good catch

Commit:
Move fetcher components to make API boundary clearer

* Move FetcherInterface to tuf/client/ directory: This way everything
  inside that directory is clearly part of client API, and everything
  outside _may_ be more of an implementation detail (settings is still
  an unfortunate exception)
* Keep RequestsFetcher in tuf/ for same reasons: it's just the default
  implementation, not explicitly part of client API

An even clearer division would be if we moved all the client specific
implementation details (download.py, mirrors.py, requests_fetcher.py)
to tuf/client/_internal/ but that's a larger change...

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
I'm using this same PR to add two changes. We've discussed these with Teodora but she has not seen the actual commits:
 * Add a custom HTTP error for Fetcher implementations to raise
 * Move the components around a bit (to make it clear what we consider client API)
 
The last point maybe deserves a definition: Through these changes we've had this idea:
* tuf.client contains the "client API"
* any other tuf.* modules are _from the client POV_ more or less implementation details

This is why there is a (minimal) module interface change in tuf.download in the PR and the last commit makes Fetcher components respect this idea as well.

There is no absolute "API boundary" defined in TUF at the moment but I think the above is a good goal to have for the client. To make this even clearer download.py, mirrors.py and requests_fetcher.py could be moved to e.g. tuf/client/_internal/ but I did not go that far in this PR.

jku commented:
Since it seems I did not report on the pip tests before this: 

I've tested this branch with pip: the custom fetcher works and is pleasant to implement, 15 LOC was enough to get pip progress bars working. From my POV this is ready (but I'm not adding an official review as I'm now an author as well).

joshuagl commented on tuf/fetcher.py:
```diff
@@ -16,6 +16,16 @@
+
+import urllib3.exceptions
+import tuf.exceptions
+import tuf.settings
```
Is this import style compatible with vendoring @jku? Or should we be doing `from foo import bar`?

jku replied:
It is not compatible but I figured this will land before the imports (and since merge fixes are going to be needed anyway I'd change these in the imports PR where I change everything else as well)

joshuagl commented on tuf/exceptions.py:
```diff
@@ -335,3 +335,12 @@ class URLParsingError(Error):
+  """
+  Returned by FetcherInterface implementations for HTTP errors.
+  status_code property is the (int) HTTP status code
+  """
```
Can we document `message` and `status_code` here?

sechkova replied:
Added some explanation in 7cd6177

joshuagl commented on tests/test_download.py:
```diff
@@ -118,18 +146,18 @@ def test_download_url_to_tempfileobj_and_lengths(self):
     # length greater than the size of the target file).
     self.assertRaises(tuf.exceptions.DownloadLengthMismatchError,
-        download.safe_download, self.url, self.target_data_length + 1)
+        download.safe_download, self.url, self.target_data_length + 1, self.fetcher)
```
should there be a `.close()` call on the downloaded file here, like the calls to `safe_download()` above and below?

joshuagl replied:
Ah, now I see that we close the file when we catch an exception. Hmm, I wonder if that's expected (anticipated) behaviour?

jku replied:
In the exception case there is no (and can't be a) file to close at this point, I think this is correct (it just looks weird in the tests since we don't ever do anything with the file -- normally you'd use a `with download.safe_download() as f:` construct)

joshuagl commented on tests/test_download.py:
```diff
@@ -164,32 +192,26 @@ def test_download_url_to_tempfileobj_and_urls(self):
+      download_file(None, self.target_data_length, self.fetcher)
 
-    url = 'http://localhost:' \
+    url = 'http://127.0.0.1:' \
```
Nit: a change like this in future might be easier if we define the local URL as a test level constant

sechkova replied:
A suggestion in 5d3133d.

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+  See LICENSE-MIT OR LICENSE for licensing information.
+
+<Purpose>
+  Unit test for 'fetcher.py'.
```
Specifically unit tests for requests_fetcher?

sechkova replied:
Fixed in be399d2.

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+import math
+
+import tuf
+import tuf.unittest_toolbox as unittest_toolbox
```
Nit: should this be sorted alphabetically? `unittest_toolbox` below `requests_fetcher`

sechkova replied:
Fixed in be399d2.

joshuagl commented on tests/test_proxy_use.py:
```diff
@@ -202,8 +205,8 @@ def test_baseline_no_proxy(self):
     logger.info('Trying HTTP download with no proxy: ' + self.url)
-    download.safe_download(self.url, self.target_data_length)
-    download.unsafe_download(self.url, self.target_data_length)
+    download.safe_download(self.url, self.target_data_length, self.fetcher)
```
do the `safe_download()` and `unsafe_download()` calls in test_proxy_use.py need to close the downloaded file?

jku replied:
Yes I think they should. For some reason there are no warnings about unclosed files on these though...

sechkova replied:
Fixed in 8df22a9.

joshuagl commented on tuf/download.py:
```diff
@@ -128,6 +111,9 @@ def unsafe_download(url, required_length):
       limit.
 
+    fetcher:
+      A FetcherInterface object that performs the network IO operations.
```
```suggestion
      An object implementing FetcherInterface that performs the network IO operations.
```

sechkova replied:
Updated in 8b58f36.

joshuagl commented on tuf/download.py:
```diff
@@ -210,174 +190,46 @@ def _download_file(url, required_length, STRICT_REQUIRED_LENGTH=True):
-        break
+  except Exception:
+    # Close 'temp_file'.  Any written data is lost.
+    temp_file.close()
```
Not strictly related to this PR, however – Is it odd that we close the file on an exception, but otherwise leave it up to the caller?

jku replied:
Not in the new code because this code is now in _download_file() which opens the temp_file and returns it: the caller would have no way of closing the file in the exception case

joshuagl reviewed (approved):
This looks really nice, great work @sechkova and @jku! I've made a few minor comments inline.

>I've tested this branch with pip: the custom fetcher works and is pleasant to implement, 15 LOC was enough to get pip progress bars working. From my POV this is ready (but I'm not adding an official review as I'm now an author as well).

15 LOC, that's quite satisfying. Really nice work all!

joshuagl commented:
Let's be sure to follow the new coding style in the new files, especially for the docstrings.

sechkova committed:
Close temp file in test_proxy_use.py

Calls to safe_download and unsafe_download leave
a temporary file unclosed.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add host address as a test level constant

Use a common test level constant for defining
the host address forming the download URL on
the client side.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Addressed all comments and suggestions (links to commits added since there was a rebase). Let me know if I am missing something or something else comes up.

MVrachev commented on tuf/download.py:
```diff
@@ -192,33 +192,33 @@ def _download_file(url, required_length, fetcher, STRICT_REQUIRED_LENGTH=True):
 
       stop_time = timeit.default_timer()
-      seconds_spent_receiving = stop_time - start_time
+      temp_file.write(chunk)
```
It seems I  am missing context, but why did you decide to move the write chunk to temp file before exiting the fetcher loop on error?

jku replied:
I believe it's because the old code can lead to an exception even if all bytes have already been downloaded (which seems a bit useless). Teodoras implementation still warns in that case but does not throw.

sechkova replied:
In both cases writing the chunk to the temp_file is in the for loop. I moved it to be before the average download speed calculation and mostly before the `break` statement on line 216 since it was resulting in misleading errors on low average download speed as Jussi has explained.

sechkova replied:
@MVrachev ah sorry, misread your question, so yeah ... what Jussi said :grin:

MVrachev commented on tuf/download.py:
```diff
@@ -192,33 +192,33 @@ def _download_file(url, required_length, fetcher, STRICT_REQUIRED_LENGTH=True):
+    start_time = timeit.default_timer()
+    for chunk in chunks:
 
       stop_time = timeit.default_timer()
```
I am wondering what exactly will happen between the `start_time` and `stop_time` given that they are just one after another.
Do `chunks` remain blocked until it's fully downloaded and then start the for loop (but that doesn't make sense, this would mean you have to have the whole file in memory which is what we want to escape)? 

Sorry for those probably obvious questions in this stage of the review process, but I wanted to make sure I understand this.

jku replied:
good question. the blocking points should be:
* `fetcher.fetch()` -- connect(), DNS lookup, TLS negotiation, etc
* each iteration of `chunks`: the iterator blocks until it has downloaded data for the chunk

this is the intention and "obvious implementation" at least: of course the Fetcher implementation could do things differently.

MVrachev replied:
> fetcher.fetch() -- connect(), DNS lookup, TLS negotiation, etc
This means that the `start_time` won't catch the delay when starting the fetch process which makes sense

> each iteration of chunks: the iterator blocks until it has downloaded data for the chunk

This means it will count the average download on each chunk which seems right.
Thanks, I think I understand it now.

joshuagl commented:
Would it be possible to ensure the new files (tuf/client/fetcher.py and tuf/requests_fetcher.py) adhere to the new coding style? Of particular note is the docstring format.

sechkova committed:
Apply the new code style to fetcher docstrings

Ensure that the newly added files' docstrings adhere to the
recently adopted code style guideline (#1232).

Small code style improvements in comments and imports.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
> Would it be possible to ensure the new files (tuf/client/fetcher.py and tuf/requests_fetcher.py) adhere to the new coding style? Of particular note is the docstring format.

The docstrings are updated. I haven't applied the rules from the new coding style which cause linter errors. Since the changes in the PR are meant to work with the existing code they follow the old style. Let's review the coding style during the transition to TUF 1.0.0

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+
+"""
+<Program>
+  test_download.py
```
Nit:
```suggestion
  test_fetcher.py
```

Maybe we should be explicit and call this test_requests_fetcher?

sechkova replied:
Fixed the docstring but test_requests_fetcher sounds too verbose :)

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+
+  # Test: Normal case.
+  def test_fetch(self):
+
```
Nit:
```suggestion
```

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+    for chunk in self.fetcher.fetch(self.url, self.file_length-4):
+        self.temp_file.write(chunk)
+
+    self.temp_file.seek(0, io.SEEK_END)
```
Nice, I didn't realise these constants were available in Python2.7.

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+  def test_url_parsing(self):
+    with self.assertRaises(tuf.exceptions.URLParsingError):
+      for chunk in self.fetcher.fetch(self.random_string(), self.file_length):
+          self.temp_file.write(chunk)
```
I assume the write never happens here? Do we need the for loop, or could we just call `self.fetcher.fetch()` to trigger the expected exception?

joshuagl commented on tests/test_fetcher.py:
```diff
@@ -0,0 +1,146 @@
+    temp_file_data = self.temp_file.read().decode('utf-8')
+    self.assertEqual(self.file_contents, temp_file_data)
+    # Double check that we download more than one chunk
+    self.assertGreater(chunks_count, 1)
```
I think this check is unnecessary? We are checking for the expected number of chunks below, what additional value does this provide?

sechkova replied:
I reworked it a bit.

Does this line express my intention better:
https://github.com/theupdateframework/tuf/blob/c794e0b917ba79327277f2846eda0c9a005b6730/tests/test_fetcher.py#L122-L123

The goal is to make sure that the calculation of `expected_chunks` is correct and results in more than one chunk downloaded.

joshuagl replied:
Oh, I had confused variables. Apologies. 🤦 
The locality of the check makes the intention clearer, I don't think we should be using `assert()` over `self.assert*` here though.

sechkova replied:
I did change the variable to make the test clearer 🙄 
Here it is, using  the unittest's "assert"  50b3b19

joshuagl commented on tuf/download.py:
```diff
@@ -192,12 +180,6 @@ def _download_file(url, required_length, STRICT_REQUIRED_LENGTH=True):
-
-  # Do all of the arguments have the appropriate format?
-  # Raise 'securesystemslib.exceptions.FormatError' if there is a mismatch.
-  securesystemslib.formats.URL_SCHEMA.check_match(url)
-  tuf.formats.LENGTH_SCHEMA.check_match(required_length)
```
Was removing the format checks for url and required_length intentional? I was half expecting them to have moved to the fetcher implementation.

sechkova replied:
It is fully intentional. The call stack is `(un)safe_download` -> `_download_file` and the same checks already exist in `(un)safe_download`

joshuagl replied:
Got it, thanks for clarifying!

joshuagl reviewed (approved):
A few minor comments, which you may choose to address or not, and a couple of questions. I will be happy to merge this in its current state once we've released v0.17.0 (#1284).

Thank you for the solid work on this!

sechkova commented:
Addressed the latest comments and rebased on the latest changes from the development branch.

jku commented:
Had another look, I agree with my previous self: LGTM.

Notes:
* The whole thing is a mostly faithful reproduction of the earlier code, just in a RequestsFetcher.fetch() generator function
* requests_fetcher.py, download.py and mirrors.py are implementation details of updater.py: this is why download.py "API change" was seen as reasonable in the PR
* download.py does not provide much value anymore but I think keeping it is the correct call to keep the PR to a reasonable size
* This PR fixes the issues that I can see with pip integration (ability to tweak network stack tweakables, ability to do progress notifications, ability to compress metadata on the wire)
* The test server address fix could be more integrated (like `TestServerProcess.get_base_url()` that returns something like "http://127.0.0.1:1234/" but I think this is good enough for this PR)

There's an approval already so I'll merge this soon if no complaints appear.

lukpueh commented:
Thanks for all the efforts you put into this PR, @sechkova and @jku! And kudos to the reviewers, @joshuagl and @MVrachev. ❤️ 

I won't add any meaningful comments here, but based on a quick glance through the diff and all the other comments and discussions above, I'm sure the approval is well justified. :)
